bip-PC.md

```diff
@@ -0,0 +1,187 @@
+<pre>
+  BIP: ?
+  Layer: Consensus (soft fork)
+  Title: OP_PAIRCOMMIT
+  Author: moonsettler <moonsettler@protonmail.com>
+  Comments-Summary: No comments yet.
+  Comments-URI: <links to wiki page for comments>
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-11-08
+  License: BSD-3-CLAUSE
+</pre>
+
+## Abstract
+
+This BIP describes a new tapscript opcode `OP_PAIRCOMMIT` which
+provide limited vector commitment functionality in tapscript.
+
+When evaluated, the `OP_PAIRCOMMIT` instruction:
+* Pops the top two values off the stack,
+* takes the "PairCommit" tagged SHA256 hash of the stack elements,
+* pushes the resulting commitment on the top of the stack.
+
+## Motivation
+
+To do LN-Symmetry contracts that don't require the nodes to keep old states,
+we need to solve the data availability problem presented by unilateral closes.
+Channel peers must be able to reconstruct the script that spends an
+intermediate state.
+
+Using in sequence `OP_CHECKTEMPLATEVERIFY`, `OP_PAIRCOMMIT`, `OP_INTERNALKEY`
+and `OP_CHECKSIGFROMSTACK` we can construct a rebindable channel that is also
+optimal.
```

Comment:
This paragraph seems to indicate that the `OP_PAIRCOMMIT` proposal would be especially useful in combination with these other opcodes. Could you perhaps clarify whether and how `OP_PAIRCOMMIT` is useful by itself in absence of the other three opcodes you mention here?

Reply:
LNhance at it's core is `CTV `+ `CSFS`. They together provide the core utility. `IKEY `is an optimization for not having to pay for the pubkey twice when the internal key can be used. `PC` is an optimization when `CSFS `has to commit to additional data required to recreate a spend script from an intermediate state, because `OP_RETURN` (to which `CTV` naturally commits to) is 4x more expensive in weight units for data availability.

`PC` could also be used by `CHECKCONTRACTVERIFY` to carry a complex state in the absence of `CAT`.

I don't think anyone would find `PC` useful enough to activate in isolation without the aforementioned other opcodes.
It can do general merkle tree style commitments that are not compatible with other merkle tree structures in bitcoin.

We probably will make a new BIP for LNhance that has these other BIPs as "Relies on".

Reply:
I meant that these explanations should be part of the BIP, not just part of the conversation here in the pull request comments.

Reply:
So should I link the "Use in LN-Symmetry" section from there? Or it needs better explanation?

Reply:
Really leaning towards making a head BIP for LNhance and keeping the individual BIPs strictly limited to describing functionality without a lot of speculation on applications. (Which we would do after finalizing the individual ops.)

Is there a problem with this approach?

Reply:
If you are creating multiple BIPs that only make sense together, it would be better to propose them as a single pull request. Since this is being proposed standalone, it should also provide its own raison d’être.

Reply:
Maybe I should write that it was an explicit design goal that PAIRCOMMIT is pretty much completely useless on it's own with the current set of opcodes?

Reply:
> I meant that these explanations should be part of the BIP, not just part of the conversation here in the pull request comments.

While BIP-2 recommends that `a single BIP contain a single key proposal or new idea. The more focused the BIP, the more successful it tends to be. If in doubt, split your BIP into several well-focused ones`, I agree that background context like https://github.com/bitcoin/bips/pull/1699#discussion_r1856937991 ought to be here in the open source BIP, and not only as GitHub meta-data (review comments) or an entry in Delving.

Reply:
PairCommit is generically useful as a concept, and I think that having things like this spec'd, implemented, and deployable is a positive good towards improving bitcoin.

Further, new things that synergize with exisiting older concepts having to go together into the same PR would create a disincentive for people to share and work on new ideas, since it would make them "less mature" as a package. For example, when I published the CTV V2 Hashing Spec WIP BIP, it would have made ZERO sense for that to go into the CTV BIP (or CTV PR, were it not merged), since it's a standalone concept. A concept that now, thanks to me writing it up, can be discussed against PairCommit as an alternative.

wholeheartedly support this being a standalone BIP and PR.

Reply:
Thanks for weighing in, @JeremyRubin.